subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/internal/precompiled/PrecompileGroovyScriptsTask.java

```diff
@@ -0,0 +1,299 @@
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.Internal;
+import org.gradle.api.tasks.OutputDirectory;
+import org.gradle.api.tasks.PathSensitive;
+import org.gradle.api.tasks.PathSensitivity;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.configuration.CompileOperationFactory;
+import org.gradle.configuration.ScriptTarget;
+import org.gradle.groovy.scripts.BasicScript;
+import org.gradle.groovy.scripts.ScriptRunner;
+import org.gradle.groovy.scripts.internal.BuildScriptData;
+import org.gradle.groovy.scripts.internal.CompileOperation;
+import org.gradle.groovy.scripts.internal.CompiledScript;
+import org.gradle.groovy.scripts.internal.ScriptCompilationHandler;
+import org.gradle.groovy.scripts.internal.ScriptRunnerFactory;
+import org.gradle.initialization.ClassLoaderScopeRegistry;
+import org.gradle.internal.Actions;
+import org.gradle.internal.classpath.DefaultClassPath;
+import org.gradle.internal.exceptions.LocationAwareException;
+import org.gradle.internal.service.ServiceRegistry;
+import org.gradle.model.dsl.internal.transform.ClosureCreationInterceptingVerifier;
+import org.gradle.plugin.management.PluginRequest;
+import org.gradle.plugin.management.internal.PluginRequests;
+import org.gradle.plugin.use.internal.PluginsAwareScript;
+import org.gradle.testfixtures.ProjectBuilder;
+
+import javax.inject.Inject;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.IOException;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@CacheableTask
+class PrecompileGroovyScriptsTask extends DefaultTask {
+
+    private final Project project = getProject();
+
+    private final ScriptCompilationHandler scriptCompilationHandler;
+    private final CompileOperationFactory compileOperationFactory;
+    private final FileSystemOperations fileSystemOperations;
+    private final ScriptRunnerFactory scriptRunnerFactory;
+    private final ServiceRegistry serviceRegistry;
+
+    private final ClassLoaderScope classLoaderScope;
+
+    private final List<PrecompiledGroovyScript> scriptPlugins;
+
+    private final Provider<Directory> intermediatePluginBlockClassesDir;
+    private final Provider<Directory> intermediatePluginClassesDir;
+    private final Provider<Directory> intermediatePluginMetadataDir;
+
+    private final DirectoryProperty precompiledGroovyScriptsOutputDir = project.getObjects().directoryProperty();
+    private final DirectoryProperty pluginAdapterSourcesOutputDir = project.getObjects().directoryProperty();
+    private final DirectoryProperty pluginAdapterClassesOutputDir = project.getObjects().directoryProperty();
+
+    private final Provider<Directory> javaSourceDependencyClasses;
+    private final Provider<Directory> groovySourceDependencyClasses;
+
+    private final FileCollection classpath;
```

Comment:
This should be a `ConfigurableFileCollection` (and it can be a managed property too)